### PR TITLE
ci: add scheduled public API e2e tests (CM-968)

### PR DESCRIPTION
**Pull request by skwowet**
## Summary

Adds a scheduled Public API e2e suite against the dedicated `api-e2e` stack (`crowd_api_e2e`), so we can catch contract regressions on `main` without hand-running against shared staging.

## Changes

- Add `.github/scripts/public-api-e2e-tests.sh` (HTTP seed + resource suites)
- Add `.github/workflows/api-e2e-tests.yml` (deploy `api-e2e` from `main`, then run the suite)
- Register `api-e2e` in `scripts/builders/backend.env`
- Document architecture and suite design in ADR-0012 / ADR-0013

### CODE REVIEW

File: .github/workflows/api-e2e-tests.yml [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+          health_url="${CDP_API_E2E_BASE_URL%/v1}/health"
+          for i in $(seq 1 36); do
+            curl -fsS --max-time 10 "$health_url" >/dev/null && exit 0
+            sleep 10
+          done
+          echo "api-e2e not healthy: $health_url" >&2
+          exit 1
+
+      - name: Run Public API e2e tests
+        run: bash .github/scripts/public-api-e2e-tests.sh
```
[the remark this answers is not in the document]
Comment (skwowet) [reply .github/workflows/api-e2e-tests.yml:80]: Fixed — added `permissions: contents: read` at the workflow level.

File: .github/workflows/api-e2e-tests.yml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+      - name: Wait for api-e2e
+        run: |
+          set -euo pipefail
+          health_url="${CDP_API_E2E_BASE_URL%/v1}/health"
+          for i in $(seq 1 36); do
+            curl -fsS --max-time 10 "$health_url" >/dev/null && exit 0
+            sleep 10
+          done
+          echo "api-e2e not healthy: $health_url" >&2
+          exit 1
```
[the remark this answers is not in the document]
Comment (skwowet) [reply .github/workflows/api-e2e-tests.yml]: Good catch — fixed. We now run `kubectl rollout status deployment/api-e2e-dpl --timeout=600s` before the public health check so we wait for the new revision, not an old pod.

File: .github/workflows/api-e2e-tests.yml [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          ref: main
+
+      - name: Install OCI
+        run: |
+          curl -L https://raw.githubusercontent.com/oracle/oci-cli/master/scripts/install/install.sh > install.sh
+          chmod +x install.sh
+          ./install.sh --accept-all-defaults
```
[the remark this answers is not in the document]
Comment (skwowet) [reply .github/workflows/api-e2e-tests.yml]: Fixed — OCI install is pinned to `v3.51.1` (not `master`), Oracle credentials are scoped to the deploy step, and Auth0 credentials are scoped to the test step only.

File: .github/workflows/api-e2e-tests.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (skwowet) [reply .github/workflows/api-e2e-tests.yml:80]: Intentional for now — Slack `#cdp-alerts` is deferred. Failures surface as a failed GitHub Actions run; we'll add the alert step in a follow-up.

File: docs/adr/0011-api-e2e-test-architecture.md [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+- **Assert**: auth/scopes, routing, validation, sync response shape and critical flows.
+- **Do not assert**: Temporal / OpenSearch eventual outcomes. Some writes await Temporal accepting a signal before the HTTP response returns; async worker work stays out of scope for this suite.
+
+### Not a PR gate
+
+A single shared `api-e2e` cannot safely gate concurrent PRs, because deploys and DB use would step on each other. This stack is a **scheduled check of what `main` deployed**. Per-PR API e2e waits on the backlog harness.
+
+### Database reset (ops)
+
+[`reset_api_e2e_test_db.sh`](../../reset_api_e2e_test_db.sh) runs on the EC2 host used to reach staging Postgres over SSH. It creates `crowd_api_e2e` if needed, applies normal Flyway migrations, truncates tables, and seeds the default tenant.
```
[the remark this answers is not in the document]
Comment (skwowet) [reply docs/adr/0011-api-e2e-test-architecture.md]: Fixed — the ADR no longer uses a repo-relative link. The reset script lives on the EC2 ops host (documented in the Database reset section), not in this checkout.

File: docs/adr/0011-api-e2e-test-architecture.md [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+## Context
+
+We need to catch regressions on the API on `main` sooner (auth, routing, validation, and the sync HTTP response contract) without waiting for a consumer to hit a break. Relying on manually running the script every time is hard and not reliable. It pollutes staging data, depends on ambient fixtures, and does not work as a team habit.
+
+A fuller API e2e harness in the server test setup (PR CI, isolated fixtures, typed asserts) is planned and sits in the backlog. Until that is ready, we need a short-term automated check that is isolated from staging data and cheap to operate.
+
+**Focus for now is the Public API.** Those endpoints are consumed outside the product, so we need a minimum unbroken contract there first. Internal API routes are not the priority for these tests; we can add coverage for them later when there is a real need, using the same stack and suite pattern.
+
+This ADR is the **runtime and isolation** decision. How e2e test cases are written lives in [ADR-0012](./0012-api-e2e-test-suite-design.md). Current suite: [`.github/scripts/public-api-e2e-tests.sh`](../../.github/scripts/public-api-e2e-tests.sh).
```
[the remark this answers is not in the document]
Comment (skwowet) [reply docs/adr/0011-api-e2e-test-architecture.md]: Updated the Context wording for clarity while keeping the background we need for this decision. Leaving a bit more than the 2–5 sentence guideline here on purpose.

File: .github/scripts/public-api-e2e-tests.sh [hunk 6]
```diff
@@ -0,0 +1,497 @@
...
+die() {
+  echo "error: $*" >&2
+  exit 1
+}
+
+api() {
+  local method=$1 path=$2 body=${3-}
+  local -a args=(-sS -w '\n%{http_code}' -X "$method" "${CDP_API_E2E_BASE_URL}${path}"
+    -H 'Content-Type: application/json'
+    -H "Authorization: Bearer ${TOKEN}")
```
[the remark this answers is not in the document]
Comment (skwowet) [reply .github/scripts/public-api-e2e-tests.sh]: Fixed — added `--connect-timeout 10` and `--max-time 30` on both `api` and `fetch_token` curls (same connect budget as the workflow health check; 30s total covers DB work + Temporal signal accept).

File: .github/scripts/public-api-e2e-tests.sh [hunk 7]
```diff
@@ -0,0 +1,497 @@
...
+PERSON_NAME="Jordan Lee"
+PERSON_LFID="jordan.lee-${RUN_ID}"
+PERSON_ID=""
+
+seed() {
+  echo "=== seed (${RUN_ID}) ==="
+
+  api POST /organizations "$(json \
+    --arg name "$ACME_NAME" --arg domain "$ACME_DOMAIN" --arg source "$SOURCE" \
+    '{name:$name, domain:$domain, source:$source}')"
```
[the remark this answers is not in the document]
Comment (skwowet) [reply .github/scripts/public-api-e2e-tests.sh]: Agreed there was a contract mismatch, but the product spec and handler treat `logo` as optional (`logo?: string` in the LFX One transactional API spec; Zod `.optional()` in `createOrganization`). OpenAPI incorrectly listed it under `required`.

Fixed OpenAPI to match the spec/handler — left the e2e suite unchanged (omitting logo is valid).

File: .github/scripts/public-api-e2e-tests.sh [hunk 8]
```diff
@@ -0,0 +1,497 @@
...
+
+  api POST "/members/${PERSON_ID}/identities" "$(json \
+    --arg email "$email" '{
+      value: $email,
+      platform: "email",
+      type: "email",
+      source: "lfxOne",
+      verified: false
+    }')"
+  check "POST same identity is idempotent" 200 ".id == \"$identity_id\""
```
[the remark this answers is not in the document]
Comment (skwowet) [reply .github/scripts/public-api-e2e-tests.sh:305]: Intentional — leaving the e2e expectation as `200`.

`createMemberIdentity` treats an exact platform+value match already on **this** member as idempotent: it sets `alreadyExisted`, returns the existing row, and responds with `200` (`ok`) instead of inserting again. A true create returns `201`. Conflicts against **another** member still surface as `409` via the DB unique path / `rethrowDbConflict`.

OpenAPI’s description (“409 if already on this member”) is the drift; the detailed API section of the product spec documents a success response for this endpoint, and the early “Identity Added → 409 same member” table disagrees with that later design. Aligning OpenAPI can be a follow-up — this suite locks the intentional runtime behavior.

File: docs/adr/0012-api-e2e-test-architecture.md [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+### What a scheduled run does
+
+1. Deploy `main` to `api-e2e`
+2. Authenticate with Auth0
+3. Seed fixtures over HTTP (`RUN_ID`-tagged names and domains)
+4. Call endpoints and assert status + body
+5. Alert `#cdp-alerts` on failure
+
+`RUN_ID` is `GITHUB_RUN_ID` in Actions, otherwise timestamp + random. Fixtures include it so runs do not collide and leftovers stay easy to spot. The suite seeds over HTTP and does not require a wiped database between runs, only a reachable `api-e2e`.
```
[the remark this answers is not in the document]
Comment (skwowet) [reply docs/adr/0012-api-e2e-test-architecture.md]: Fixed — ADR now describes `RUN_ID` as `GITHUB_RUN_ID` (or a local timestamp) plus a `${RANDOM}` suffix, noting that workflow reruns reuse the same run ID.

File: .github/scripts/public-api-e2e-tests.sh [hunk 10]
```diff
@@ -0,0 +1,502 @@
...
+
+CURL_CONNECT_TIMEOUT=10
+CURL_MAX_TIME=30
+
+api() {
+  local method=$1 path=$2 body=${3-}
+  local -a args=(-sS --connect-timeout "$CURL_CONNECT_TIMEOUT" --max-time "$CURL_MAX_TIME"
+    -w '\n%{http_code}' -X "$method" "${CDP_API_E2E_BASE_URL}${path}"
+    -H 'Content-Type: application/json'
+    -H "Authorization: Bearer ${TOKEN}")
```
[the remark this answers is not in the document]
Comment (skwowet) [reply .github/scripts/public-api-e2e-tests.sh:67]: Fair point — this suite is intentionally happy-path only. A broken token/client still fails the run; missing-token 401 / insufficient-scope 403 (auth becoming too permissive) is out of scope for now.

Narrowed ADR-0012 so we no longer claim auth/scope rejection coverage: we assert routing, validation, and the sync contract using a valid fully scoped token.

File: docs/adr/0013-api-e2e-test-suite-design.md [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+
+We want a thin, boring pattern: shared helpers, clear grouping, and an obvious place to add the next resource. This is how we write durable e2e cases until a fuller PR-time API e2e harness exists (backlog).
+
+This ADR is the **suite design** (how tests are written). Runtime, isolation, scheduled CI, which surfaces we cover, and ops live in [ADR-0012](./0012-api-e2e-test-architecture.md).
+
+## Decision
+
+Implement API e2e tests as HTTP bash script(s) with this structure. Current entrypoint: [`.github/scripts/public-api-e2e-tests.sh`](../../.github/scripts/public-api-e2e-tests.sh). New suites should reuse this pattern (same script or additional scripts), not invent a parallel style.
+
+The suite is **thin**: assert the HTTP contract and critical flows so regressions show up early. Leave exhaustive edge-case matrices to unit or focused contract tests.
```
[the remark this answers is not in the document]
Comment (skwowet) [reply docs/adr/0013-api-e2e-test-suite-design.md:19]: I think that's fine.

File: .claude/skills/write-unit-tests/SKILL.md [hunk 12]
```diff
@@ -0,0 +1,71 @@
+---
+name: write-unit-tests
+description: >
+  Write focused Vitest unit tests. Use when adding or improving unit tests for
+  business logic, data access, common services, or other server modules.
```
[the remark this answers is not in the document]
Comment (skwowet) [reply .claude/skills/write-unit-tests/SKILL.md]: Yeah, moved the skills to different pr

File: .github/scripts/public-api-e2e-tests.sh [hunk 13]
```diff
@@ -0,0 +1,502 @@
...
+}
+
+json() {
+  jq -n "$@"
+}
+
+fetch_token() {
+  local response
+  response="$(curl -sS --connect-timeout "$CURL_CONNECT_TIMEOUT" --max-time "$CURL_MAX_TIME" \
+    -X POST "${AUTH0_STAGING_ISSUER}/oauth/token" \
```
[the remark this answers is not in the document]
Comment (skwowet) [reply .github/scripts/public-api-e2e-tests.sh]: Fixed — strip a trailing slash from `AUTH0_STAGING_ISSUER` before appending `/oauth/token` so `https://…/` does not become `//oauth/token`.

File: docs/adr/0008-how-we-write-unit-tests.md [hunk 14]
```diff
@@ -25,6 +25,8 @@ reuse is real.
 - Setup should make that story obvious (concrete labels, dates, relationships).
 - Do not combine unrelated behaviors in a single test.
 - Prefer fewer tests that fail clearly over many that pass for the wrong reason.
+- When a file covers more than one function, group cases with
+  `describe('<functionName>', …)` and put named scenarios inside.
```
[the remark this answers is not in the document]
Comment (skwowet) [reply docs/adr/0008-how-we-write-unit-tests.md]: Moved — ADR-0008 mocking/grouping edits are now on #4402 with the testing skills.

File: .github/workflows/api-e2e-tests.yml [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+          ORACLE_DOCKER_USERNAME: ${{ secrets.ORACLE_DOCKER_USERNAME }}
+          ORACLE_DOCKER_PASSWORD: ${{ secrets.ORACLE_DOCKER_PASSWORD }}
+          ORACLE_USER: ${{ secrets.ORACLE_USER }}
+          ORACLE_TENANT: ${{ secrets.ORACLE_TENANT }}
+          ORACLE_REGION: ${{ secrets.ORACLE_REGION }}
+          ORACLE_FINGERPRINT: ${{ secrets.ORACLE_FINGERPRINT }}
+          ORACLE_KEY: ${{ secrets.ORACLE_KEY }}
+          ORACLE_CLUSTER: ${{ secrets.ORACLE_STAGING_CLUSTER }}
+        with:
+          services: api-e2e
```
[the remark this answers is not in the document]
Comment (skwowet) [reply .github/workflows/api-e2e-tests.yml:58]: Fixed — after checking out `main`, we set the builder `tag` from `git rev-parse HEAD` so the image tag matches the deployed checkout (not a possibly-mismatched `github.sha`).

File: .github/scripts/public-api-e2e-tests.sh [hunk 16]
```diff
@@ -0,0 +1,503 @@
...
+        value: $lfid,
+        platform: "lfid",
+        type: "username",
+        source: "lfxOne",
+        verified: true,
+        verifiedBy: $by
+      }]
+    }')"
+  check "POST create Sam" 201 'has("memberId")'
+  sam_id="$(jq -r '.memberId' <<<"$BODY")"
```
[the remark this answers is not in the document]
Comment (skwowet) [reply .github/scripts/public-api-e2e-tests.sh]: Fixed — added `body_field` to soft-read IDs from the response body after soft `check`s, so empty/invalid JSON no longer aborts the suite under `set -e`.

File: docs/adr/README.md [hunk 17]
```diff
@@ -18,6 +18,8 @@ Use the `/adr` skill in Claude Code to record new ADRs or query past decisions.
 | [ADR-0009](./0009-packagist-worker-design-decisions.md)        | Packagist worker — design decisions                                                                  | accepted | 2026-07-13 |
 | [ADR-0010](./0010-security-contacts-worker.md)                 | Security contacts — tiered extraction, confidence scoring, and Temporal batch ingestion              | accepted | 2026-07-21 |
 | [ADR-0011](./0011-mailinglist-skip-unparseable-dates.md)       | Skip mailing list activities with unparseable/implausible dates                                      | accepted | 2026-07-19 |
+| [ADR-0012](./0012-api-e2e-test-architecture.md)                | API e2e test architecture                                                                            | accepted | 2026-07-25 |
+| [ADR-0013](./0013-api-e2e-test-suite-design.md)                | API e2e test suite design                                                                            | accepted | 2026-07-24 |
```
[the remark this answers is not in the document]
Comment (skwowet) [reply docs/adr/README.md:22]: Fixed — PR description now references ADR-0012 / ADR-0013.

File: .github/workflows/api-e2e-tests.yml [hunk 18]
```diff
@@ -0,0 +1,114 @@
+name: API e2e tests
+
+on:
+  schedule:
+    - cron: '0 1 * * *'
+  workflow_dispatch:
+
+concurrency:
+  group: api-e2e-tests
+  cancel-in-progress: false
```
Comment (ulemons) [inline .github/workflows/api-e2e-tests.yml]: might have sense to have this true ? is this intentional ?

**Review by ulemons — DISMISSED** [review]
LGTM ✅ 

Great job!

Do you think it could be possible/good idea to have fixtures in static file outside of the `tests.sh` files? just curious

File: .github/workflows/api-e2e-tests.yml (hunk 18 above)
Comment (skwowet) [reply .github/workflows/api-e2e-tests.yml]: Yeah makes sense — will flip to true. Especially useful for manual re-dispatch so we don’t wait on a stuck/old run. Either way is low impact for a short midnight job, but true is cleaner here.

**Comment by skwowet** [comment]
> Do you think it could be possible/good idea to have fixtures in static file outside of the tests.sh files?

@ulemons Yeah, that's doable — the request payloads could live in JSON files and the script just reads them.

Might do that if the suite grows; right now most of them are small and carry a `RUN_ID` so they don't collide between runs, so inline felt simpler.
